### ISSUE TITLE
Add support for point cloud, e.g., pcd file or point cloud object from PCL

### ISSUE THREAD

**Issue by shelper** [napari/napari#1466]
## 🚀 Feature

Besides the volumetric data for 3D images, there is also a point cloud data, that is often used in 3D depth sensors like lidar, or stereo camera. 

## Motivation
3D imaging technologies like lidar, TOF camera, stereo vision camera, have become the booming area. 
point cloud based object detection/ segmentation also needs a tool like napari for visualization and annotation, etc. 

## Pitch
1. support loading and visualization of .pcd file
2. support changing camere perspective like 3d volumetric data
2. support manual annotation and segmentation of point cloud

**Comment by tlambert03** [comment]
Hi @shelper, and thanks for the suggestion!

Do you have python experience?  This seems like a good fit as a napari plugin wrapping, e.g. [`pypcd`](https://github.com/dimatura/pypcd) (or do you have another package that you recommend for reading a .pcd file into a numpy array?), to read a pcd file into a Points layer.  If you'd like to take a stab at it, here are the docs for [how to write a napari plugin](https://napari.org/docs/plugins/for_plugin_developers.html), and a [template package to get started](https://github.com/napari/cookiecutter-napari-plugin).

If it's not something you have experience with, someone here might be able to get this up and running

**Comment by shelper** [comment]
@tlambert03 sounds good, I will find sometime to take a look, to see if i can help
thanks!

**Comment by shelper** [comment]
I spent some time at the documents and don't think the point layer is a
good Choice for the point cloud...

Point layer is more or less for marking points on the image that are very
sparse

Point cloud could be millions of points which does not have size or shape
but just a point in 3D

A better choice would be surface layer but that needs to convert points
cloud to surface by triangulation of all the points...

I wonder if we should add a new layer type for point cloud

I beliece vispy support point cloud

Talley Lambert <notifications@github.com> 于 2020年7月16日周四 下午5:17写道：

**Comment by tlambert03** [comment]
> I spent some time at the documents and don't think the point layer is a good Choice for the point cloud... I believe vispy support point cloud

are you sure?  Here's an example of someone using vispy's `Markers` Visual to plot a dense point cloud https://github.com/vispy/vispy/issues/1496 which is exactly what our Points layer is backed by... I believe @jni has done stuff with millions of points with our Points layer as well.  I would give it a shot and then optimize if we find that it's not performing well.

In any case, if you happen to have an example pcd file that you could post here, I'd be curious to see one

**Comment by shelper** [comment]
@tlambert03 sorry, my mistake, point layer should work. will spend some time for it, thanks!

**Comment by tlambert03** [comment]
Hey @shelper, here's something to start from, copy this code to a file and name it `napari_pcd.py` and put it somewhere in your python path.  This is what a plugin looks like:

```python
from napari_plugin_engine import napari_hook_implementation

@napari_hook_implementation
def napari_get_reader(path):
    if path.endswith(".pcd"):
        return pcd_reader

def pcd_reader(path):
    import open3d as o3d
    import numpy as np

    pcd = o3d.io.read_point_cloud(path)
    return [(np.asarray(pcd.points), {'size': 1}, 'points')]
```

then `pip install open3d`, then run napari and drop a pcd file onto the viewer:
here's what I got (1,366,408 points)... plenty of things that might be changed about rendering parameters, or reading in metadata, and we can probably identify some things that would improve performance... but it's at least a proof of principle:
![Untitled](https://user-images.githubusercontent.com/1609449/87841527-ec54dc00-c873-11ea-884f-88542d8123b9.gif)

... and the plugin could also optionally do the triangulation to turn it into a surface if that's what you wanted

**Comment by liu4lin** [comment]
@shelper For those expecting lidar annotation features, you may check it out : https://github.com/basicai/xtreme1

**Comment by psobolewskiPhD** [comment]
LIDAR is super cool, but I think the solution here is a reader plugin for those files?

**Comment by psobolewskiPhD** [comment]
Yeah I don't think there is a napari issue here really, just a question of a reader plugin.
A bit more info here with a script
https://forum.image.sc/t/visualize-point-cloud-sequences-and-other-pointcloud-features-in-napari/84848
plus maybe relevant for anyone wanting to make a reader:
https://github.com/daavoo/pyntcloud
